kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/utils/TestUtils.scala

```diff
@@ -0,0 +1,244 @@
+    readSchema: StructType = null,
+    tableClient: TableClient = defaultTableClient): Seq[Row] = {
+
+    val result = ArrayBuffer[Row]()
+
+    var scanBuilder = snapshot.getScanBuilder(tableClient)
+
+    if (readSchema != null) {
+      scanBuilder = scanBuilder.withReadSchema(tableClient, readSchema)
+    }
+
+    val scan = scanBuilder.build()
+
+    val scanState = scan.getScanState(tableClient);
+    val fileIter = scan.getScanFiles(tableClient)
+    // TODO serialize scan state and scan rows
+
+    fileIter.forEach { fileColumnarBatch =>
+      // TODO deserialize scan state and scan rows
+      val dataBatches = Scan.readData(
+        tableClient,
+        scanState,
+        fileColumnarBatch.getRows(),
+        Optional.empty()
+      )
+
+      dataBatches.forEach { batch =>
+        val selectionVector = batch.getSelectionVector()
+        val data = batch.getData()
+
+        var i = 0
+        val rowIter = data.getRows()
+        try {
+          while (rowIter.hasNext) {
+            val row = rowIter.next()
+            if (!selectionVector.isPresent || selectionVector.get.getBoolean(i)) { // row is valid
+              result.append(row)
+            }
+            i += 1
+          }
+        } finally {
+          rowIter.close()
+        }
+      }
+    }
+    result
+  }
+
+  def withTimeZone(zoneId: String)(f: => Unit): Unit = {
+    val currentDefault = TimeZone.getDefault
+    try {
+      TimeZone.setDefault(TimeZone.getTimeZone(zoneId))
+      f
+    } finally {
+      TimeZone.setDefault(currentDefault)
+    }
+  }
+
+  def checkTable(
+    path: String,
+    expectedAnswer: Seq[TestRow],
+    readCols: Seq[String] = null,
+    tableClient: TableClient = defaultTableClient,
+    expectedSchema: StructType = null
+    // filter
+    // version
+  ): Unit = {
+
+    val snapshot = Table.forPath(path).getLatestSnapshot(tableClient)
+
+    val readSchema = if (readCols == null) {
+      null
+    } else {
+      val schema = snapshot.getSchema(tableClient)
+      new StructType(readCols.map(schema.get(_)).asJava)
+    }
+
+    if (expectedSchema != null) {
+      assert(
+        expectedSchema == snapshot.getSchema(tableClient),
+        s"""
+           |Expected schema does not match actual schema:
+           |Expected schema: $expectedSchema
+           |Actual schema: ${snapshot.getSchema(tableClient)}
+           |""".stripMargin
+      )
+    }
+
+    val result = readSnapshot(snapshot, readSchema = readSchema, tableClient = tableClient)
+    checkAnswer(result, expectedAnswer)
+  }
+
+  def checkAnswer(result: => Seq[Row], expectedAnswer: Seq[TestRow]): Unit = {
+    checkAnswer(result.map(TestRow(_)), expectedAnswer)
+  }
+
+  def checkAnswer(result: Seq[TestRow], expectedAnswer: Seq[TestRow]): Unit = {
+    if (!compare(prepareAnswer(result), prepareAnswer(expectedAnswer))) {
+      fail(genErrorMessage(expectedAnswer, result))
+    }
+  }
+
+  def prepareAnswer(answer: Seq[TestRow]): Seq[TestRow] = {
```

Comment:
do we expect test code to user this? can this be private?

Reply:
Nope good idea updated